src/BaseListView.php

```diff
@@ -420,7 +428,36 @@ private function renderPagination(): string
             return '';
         }
 
-        return $dataReader->isPaginationRequired() ? $this->pagination : '';
+        if (!$dataReader->isPaginationRequired()) {
+            return '';
+        }
+
+        if (is_string($this->pagination)) {
+            return $this->pagination;
+        }
+
+        $pagination = $this->pagination ??
+            $dataReader instanceof KeysetPaginator ? KeysetPagination::widget() : OffsetPagination::widget();
```

Comment:
Should we check for `OffsetableDataInterface` and `LimitableDataInterface`?

Reply:
done